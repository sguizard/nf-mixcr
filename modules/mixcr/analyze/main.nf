process MIXCR_ANALYZE {
    tag "$meta.id"

    container 'ghcr.io/milaboratory/mixcr/mixcr:4.6.0'
    // container 'https://github.com/sguizard/mixcr-singularity-container/blob/master/mixcr.sif'

    input:
    tuple val(meta), path(r1), path(r2)
    val(preset)
    path(library)
    path(license)

    output:
    tuple val(meta), path("*.align.report.json")   , emit: align_report_json
    tuple val(meta), path("*.align.report.txt")    , emit: align_report_txt
    tuple val(meta), path("*.assemble.report.json"), emit: assemble_report_json
    tuple val(meta), path("*.assemble.report.txt") , emit: assemble_report_txt
    tuple val(meta), path("*.clns")                , emit: clns
    tuple val(meta), path("*.clones_TRB.tsv")      , emit: clones_TRB_tsv
    tuple val(meta), path("*.qc.json")             , emit: qc_json
    tuple val(meta), path("*.qc.txt")              , emit: qc_txt
    tuple val(meta), path("*_non_refined.vdjca")   , emit: vdjca
    tuple val(meta), path("*.refined.vdjca")       , emit: refined_vdjca
    tuple val(meta), path("*.refine.report.json")  , emit: refine_report_json
    tuple val(meta), path("*.refine.report.txt")   , emit: refine_report_txt
    tuple val(meta), path("*.log")                 , emit: log
    path  "versions.yml"                           , emit: versions

    script:
    def args   = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: meta.id
    def lib    = library.toString() == 'NO_FILE' ? '' : "--library ${library}"
    lib = lib.replaceFirst('.json.gz', '')

    """
    export MI_LICENSE_FILE="$PWD/${license}"

    mixcr analyze \\
        $preset \\
        $args \\
        $lib \\
        --threads ${task.cpus} \\
        $r1 \\
        $r2 \\
        ${prefix} > ${prefix}.log

    mv ${prefix}.vdjca ${prefix}_non_refined.vdjca

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        MiXCR: \$( mixcr --version | grep -e 'MiXCR' | sed 's/MiXCR v//' | sed 's/ .\\+//' )
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: id
    """
    touch ${prefix}.align.report.json
    touch ${prefix}.align.report.txt
    touch ${prefix}.assemble.report.json
    touch ${prefix}.assemble.report.txt
    touch ${prefix}.clns
    touch ${prefix}.clones_TRB.tsv
    touch ${prefix}.qc.json
    touch ${prefix}.qc.txt
    touch ${prefix}.refined.vdjca
    touch ${prefix}.refine.report.json
    touch ${prefix}.refine.report.txt
    touch ${prefix}_non_refined.vdjca

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        MiXCR: \$( mixcr --version | grep -e 'MiXCR' | sed 's/MiXCR v//' | sed 's/ .\\+//' )
    END_VERSIONS
    """
}
