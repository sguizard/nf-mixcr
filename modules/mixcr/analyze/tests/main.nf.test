nextflow_process {

    name "Test Process MIXCR_ANALYZE"
    script "../main.nf"
    process "MIXCR_ANALYZE"
    config "./nextflow.config"

    tag "mixcr"
    tag "mixcr/analyze"

    test("MIXCR_ANALYSIS generic-amplicon-with-umi") {

        when {
            process {
                """
                // define inputs of the process here. Example:
                input[0] =
                [
                    ['id':'TCR_cat'],
                    [
                        file("${projectDir}/test-datasets/23570MA0002L01_1_20k.fastq.gz"),
                        file("${projectDir}/test-datasets/23570MA0002L01_2_20k.fastq.gz")
                    ]
                ]
                input[1] = 'generic-amplicon-with-umi'
                input[2] = file("${projectDir}/test-datasets/imgt.202312-3.sv8.json.gz")
                """
            }
        }

        then {
            def lines     = new File(process.out.log.get(0).get(1)).readLines()
            def lastLine  = lines.get(lines.size()-1)

            assert lastLine == 'Analysis finished successfully.'
            assert snapshot(process.out.align_report_json).match("align_report_json")
            assert snapshot(process.out.assemble_report_json).match("assemble_report_json")
            assert snapshot(process.out.clns).match("clns")
            assert snapshot(process.out.clones_TRB_tsv).match("clones_TRB_tsv")
            assert snapshot(process.out.qc_json).match("qc_json")
            assert snapshot(process.out.qc_txt).match("qc_txt")
            assert snapshot(process.out.refined_vdjca).match("refined_vdjca")
            assert snapshot(process.out.refine_report_json).match("refine_report_json")
            assert snapshot(process.out.vdjca).match("vdjca")
            assert snapshot(process.out.versions).match("versions")
        }
    }

}

