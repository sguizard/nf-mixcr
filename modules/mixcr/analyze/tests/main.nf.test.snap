{
    "clns": {
        "content": [
            [
                [
                    {
                        "id": "TCR_cat"
                    },
                    "TCR_cat.clns:md5,51df18d9a7af225237996eb4e5a12453"
                ]
            ]
        ],
        "timestamp": "2024-01-19T22:53:00.19639"
    },
    "qc_txt": {
        "content": [
            [
                [
                    {
                        "id": "TCR_cat"
                    },
                    "TCR_cat.qc.txt:md5,ad6c525b66ff49fb038c3af03ca033ba"
                ]
            ]
        ],
        "timestamp": "2024-01-19T22:53:00.553265"
    },
    "clones_TRB_tsv": {
        "content": [
            [
                [
                    {
                        "id": "TCR_cat"
                    },
                    "TCR_cat.clones_TRB.tsv:md5,cd23934a15f41111c0e1bd580bfbed4d"
                ]
            ]
        ],
        "timestamp": "2024-01-19T22:53:00.296818"
    },
    "versions": {
        "content": [
            [
                "versions.yml:md5,e803181bdeacebae68ee9a0c5d40ae5f"
            ]
        ],
        "timestamp": "2024-01-19T22:53:01.820397"
    },
    "refined_vdjca": {
        "content": [
            [
                [
                    {
                        "id": "TCR_cat"
                    },
                    "TCR_cat.refined.vdjca:md5,56924a5acbc384af0bdf4a224dc749c8"
                ]
            ]
        ],
        "timestamp": "2024-01-19T22:53:00.69885"
    },
    "vdjca": {
        "content": [
            [
                [
                    {
                        "id": "TCR_cat"
                    },
                    "TCR_cat_non_refined.vdjca:md5,79ec9f9d0e8ec7154c30f1e5638eb5ae"
                ]
            ]
        ],
        "timestamp": "2024-01-19T22:53:01.369715"
    },
    "qc_json": {
        "content": [
            [
                [
                    {
                        "id": "TCR_cat"
                    },
                    "TCR_cat.qc.json:md5,429f9adc170ee40d1de86cb15d122a0f"
                ]
            ]
        ],
        "timestamp": "2024-01-19T22:53:00.42199"
    },
    "align_report_json": {
        "content": [
            [
                [
                    {
                        "id": "TCR_cat"
                    },
                    "TCR_cat.align.report.json:md5,de0fe5e181784dfe4b9fe6ca6e3d4e3c"
                ]
            ]
        ],
        "timestamp": "2024-01-19T22:52:59.976085"
    },
    "assemble_report_json": {
        "content": [
            [
                [
                    {
                        "id": "TCR_cat"
                    },
                    "TCR_cat.assemble.report.json:md5,67ab3fa3f67f250fe8ff4425b22d73f9"
                ]
            ]
        ],
        "timestamp": "2024-01-19T22:53:00.118448"
    },
    "refine_report_json": {
        "content": [
            [
                [
                    {
                        "id": "TCR_cat"
                    },
                    "TCR_cat.refine.report.json:md5,8d87a511de01ed5d06ca8c679b7c774b"
                ]
            ]
        ],
        "timestamp": "2024-01-19T22:53:01.096971"
    }
}